### ISSUE TITLE
Training on Multiple GPUs

### ISSUE THREAD

**Issue by saqib-anwar** [j96w/DenseFusion#32]
Hello, thanks for sharing the code
I am new to PyTorch. As far as I understand, the code provided/ DataParallel function should run training(Posenet) on multiple gpus however, in my case it is using only one out of four gpus. I have a gpu cluster of 4x 1080ti. Also, torch.cuda.device_count() shows 4. Could you please tell me if you have used multiple gpus for training and is there something I am missing here?

**Comment by saqib-anwar** [comment]
> I think maybe you can take a try on modifying the code `export CUDA_VISIBLE_DEVICES=0` in `experiments/scripts/train_XXX.sh` into using all the GPUs. Also, I think the pspnet has already used `nn.DataParallel`, see `lib/network.py`.

I have it already modified; export CUDA_VISIBLE_DEVICES=0,1,2,3 and yes pspnet has already used nn.DataParallel; still it uses only one gpu for training

**Comment by saqib-anwar** [comment]
Could you please share if you have used this code to train on multiple gpus?

**Comment by HeikoEngemann** [comment]
Hello, 
thank you very much for sharing your project. I am facing the same issue. I am wondering if it is possible to train the whole "pipeline" in parallel. Did you try it? If you could provide a small "How to", that would be nice. I tried already all the standard stuff and I am stuck at the moment.

**Comment by kaixin-bai** [comment]
parser.add_argument('--batch_size', type=int, default=8, help='batch size')
the code above it the definition of 'batch_size', but it's not used to load data, see below that batch_size is always '1'.
dataloader = torch.utils.data.DataLoader(dataset, batch_size=1, shuffle=True, num_workers=opt.workers)

i have tried to train densefusion with multiple gpus, but i found that i must rewrite the network and loss to fit the new parameter.

**Comment by saqib-anwar** [comment]
@kaixin-bai Hello, have you had success with training on multiple gpus after rewriting the network and loss or you are still working on that at the moment?

**Comment by kaixin-bai** [comment]
@saqib-anwar still working on that

**Comment by kaixin-bai** [comment]
@saqib-anwar https://github.com/j96w/DenseFusion/issues/16  j96w has explained the problem of batch size. i tried with multiple gpus, but found that there's no difference, not be faster, and now i changed the input and network of densefusion, still working on that.

**Comment by kaixin-bai** [comment]
@saqib-anwar 
```
    if not refine:
        if idx[0].item() in sym_list:  
```
in loss.py, the above part is for symmetric object, but idx[0] gives up the other objects in idx if the batch size in idx is not 1.
and you should also change` out_rx, out_tx, out_cx` in network.py

**Comment by BacheryChan** [comment]
@saqib-anwar @kaixin-bai Hello, have you solved the problem of training on multiple gpus? The training on my machine cannot use up one single gpu's memory and it is extremely slow. Did you meet the same problem?
Please tell me if you have any solution, thanks!

**Comment by 704051419** [comment]
@kaixin-bai 你好 你用多卡可以训练了么 可以分享一下么

**Comment by SheehanChen** [comment]
> @saqib-anwar @kaixin-bai Hello, have you solved the problem of training on multiple gpus? The training on my machine cannot use up one single gpu's memory and it is extremely slow. Did you meet the same problem? Please tell me if you have any solution, thanks!
i have the same question and it take only 2GB GPU memory of my 1080.
